src/DataLoader.Uno/DataLoaderViewController.cs

```diff
@@ -0,0 +1,407 @@
+				RefreshVisualState(view, ErrorVisualGroup);
+				RefreshVisualState(view, LoadingVisualGroup);
+				var combinedVisualState = RefreshVisualState(view, CombinedVisualGroup);
+
+				_isVisualStateRefreshRequired = false;
+
+				if (_logger.IsEnabled(LogLevel.Information))
+				{
+					_logger.LogInformation($"Updated VisualStates to '{combinedVisualState}'.");
+				}
+			}
+		}
+
+		private static string GetDataVisualState(IDataLoaderState state)
+		{
+			if (state.IsInitial)
+			{
+				return InitialVisualStateName;
+			}
+			else if (state.IsEmpty)
+			{
+				return EmptyVisualStateName;
+			}
+			else
+			{
+				return DataVisualStateName;
+			}
+		}
+
+		private static string GetLoadingStateVisualState(IDataLoaderState state)
+		{
+			return state.IsLoading
+				? LoadingVisualStateName
+				: NotLoadingVisualStateName;
+		}
+
+		private static string GetErrorVisualState(IDataLoaderState state)
+		{
+			return state.Error != null
+				? ErrorVisualStateName
+				: NoErrorVisualStateName;
+		}
+
+		private DataLoaderViewState CreateDataLoaderViewState(IDataLoaderState state, Control view)
+		{
+			return new DataLoaderViewState()
+			{
+				View = view,
+				Parent = _delegate.GetDataContext(),
+				Source = _dataLoader,
+				Request = state.Request,
+				Data = state.Data,
+				Error = state.Error,
+				IsLoading = state.IsLoading,
+				IsInitial = state.IsInitial,
+				IsEmpty = state.IsEmpty
+			};
+		}
+
+		private async Task RunOnDispatcher(Action action)
+		{
+			if (Dispatcher.HasThreadAccess)
+			{
+				action();
+			}
+			else
+			{
+				await Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () => action());
+			}
+		}
+
+		// This is an optimization to avoid calling VisualStateManager.GoToState if we're already in the right state.
+		// It's also used to be able to save the visual states for each group when the control isn't loaded yet and restore them later.
+		private Dictionary<string, string> _currentGroupStates = new Dictionary<string, string>();
+
+		private void GoToState(Control control, string group, string visualState, bool useTransitions = true)
+		{
+			if (_currentGroupStates.TryGetValue(group, out var currentState))
+			{
+				if (currentState != visualState)
+				{
+					_currentGroupStates[group] = visualState;
+					if (_isLoaded)
+					{
+						VisualStateManager.GoToState(control, visualState, useTransitions);
+					}
+				}
+			}
+			else
+			{
+				_currentGroupStates[group] = visualState;
+				if (_isLoaded)
+				{
+					VisualStateManager.GoToState(control, visualState, useTransitions);
+				}
+			}
+		}
+
+		private string RefreshVisualState(Control control, string group, bool useTransitions = true)
+		{
+			if (_currentGroupStates.TryGetValue(group, out var currentState))
+			{
+				VisualStateManager.GoToState(control, currentState, useTransitions);
+				return currentState;
+			}
+			else
+			{
+				return null;
+			}
+		}
+
+		public async Task<IDisposable> LockAsync()
+		{
```

Comment:
Maybe it's a good opportunity to use AsyncLock instead of this

Reply:
no we want to avoid Uno.Core types